File: 2379.minimum-recolors-to-get-k-consecutive-black-blocks.py
#
# @lc app=leetcode id=2379 lang=python3
#
# [2379] Minimum Recolors to Get K Consecutive Black Blocks
#

# @lc code=start
class Solution:
    def minimumRecolors(self, blocks: str, k: int) -> int:
        # change = 0
        # change_record = []
        # for i, c  in enumerate(blocks):
        #     #Enter the window
        #     if c == "W":
        #         change += 1
        #     if i < k - 1:
        #         continue
        #     change_record.append(change)
        #     #Exit the window
        #     if blocks[i - k + 1] == "W":
        #         change -= 1   
        # return min(change_record) if change_record else 0    

        cnt, ans = 0, float('inf')
        for i, e in enumerate(blocks):
            cnt += 1 if e == 'W' else 0
            if i < k - 1: 
                continue
            ans = min(ans, cnt)
            cnt -= 1 if blocks[i - k + 1] == 'W' else 0
        return ans if ans else 0
# @lc code=end

File: 2379.minimum-recolors-to-get-k-consecutive-black-blocks.java
/*
 * @lc app=leetcode id=2379 lang=java
 *
 * [2379] Minimum Recolors to Get K Consecutive Black Blocks
 */

// @lc code=start
class Solution {
    public int minimumRecolors(String blocks, int k) {
        int result = Integer.MAX_VALUE;
        int n = blocks.length();
        int black = 0;
        for(int i = 0; i < n; i++){
            if(blocks.charAt(i) == 'B'){
                black++;
            }
            if(i >= k - 1){
                result = Math.min(result, k - black);
                if(blocks.charAt(i - k + 1) == 'B'){
                    black--;
                }

            }
        }
        return result;
    }
}
// @lc code=end

